hedera-node/src/main/java/com/hedera/services/store/schedule/HederaScheduleStore.java

```diff
@@ -183,49 +251,245 @@ public boolean isCreationPending() {
 		}
 
 		return Pair.of(null, schedule);
 	}
 
+	@Override
+	public ResponseCodeEnum preMarkAsExecuted(ScheduleID id) {
+		var status = usabilityCheck(id, false, null);
+		if (status != OK) {
+			return status;
+		}
+		return OK;
+	}
+
 	@Override
 	public ResponseCodeEnum markAsExecuted(ScheduleID id, Instant consensusTime) {
-		var status = usabilityCheck(id, false);
+		var status = usabilityCheck(id, false, null);
 		if (status != OK) {
 			return status;
 		}
-		final var offset = properties.triggerTxnWindBackNanos();
-		final var executionTime = consensusTime.plusNanos(offset);
-		apply(id, schedule -> schedule.markExecuted(executionTime));
+		apply(id, schedule -> schedule.markExecuted(consensusTime));
 		return OK;
 	}
 
 	@Override
-	public void expire(EntityId entityId) {
-		var id = entityId.toGrpcScheduleId();
+	public void expire(ScheduleID id) {
 		if (id.equals(pendingId)) {
 			throw new IllegalArgumentException(String.format(
 					"Argument 'id=%s' refers to a pending creation!",
 					readableId(id)));
 		}
-		var schedule = get(id);
-		schedules.get().remove(fromLong(entityId.num()));
-		extantSchedules.remove(schedule);
+
+		throwIfMissing(id);
+
+
+		var idToDelete = new EntityNumVirtualKey(fromScheduleId(id));
+		var existingSchedule = schedules.get().byId().remove(idToDelete);
+
+		if (existingSchedule != null) {
+
+			var secondKey = new SecondSinceEpocVirtualKey(existingSchedule.calculatedExpirationTime().getSeconds());
+
+			var bySecond = schedules.get().byExpirationSecond().get(secondKey);
+
+			if (bySecond != null) {
+				bySecond = bySecond.asWritable();
+				bySecond.removeId(existingSchedule.calculatedExpirationTime(), idToDelete.getKeyAsLong());
+
+				if (bySecond.getIds().isEmpty()) {
+					schedules.get().byExpirationSecond().remove(secondKey);
+				} else {
+					schedules.get().byExpirationSecond().put(secondKey, bySecond);
+				}
+			}
+
+
+			var equalityKey = new ScheduleEqualityVirtualKey(existingSchedule.equalityCheckKey());
+
+			var byEquality = schedules.get().byEquality().get(equalityKey);
+
+			if (byEquality != null) {
+				byEquality = byEquality.asWritable();
+				byEquality.remove(existingSchedule.equalityCheckValue(), idToDelete.getKeyAsLong());
+
+				if (byEquality.getIds().isEmpty()) {
+					schedules.get().byEquality().remove(equalityKey);
+				} else {
+					schedules.get().byEquality().put(equalityKey, byEquality);
+				}
+			}
+		}
+	}
+
+	@Override
+	public boolean advanceCurrentMinSecond(Instant consensusTime) {
+
+		boolean changed = false;
+
+		long curSecond = schedules.get().getCurrentMinSecond();
+
+		while ((curSecond < Long.MAX_VALUE) && ((curSecond + 1) < Instant.MAX.getEpochSecond()) &&
+				// for processing, we only use the second component for comparison, we only process stuff from the
+				// _previous_ second
+				(consensusTime.getEpochSecond() > (curSecond + 1))
+				&& (!schedules.get().byExpirationSecond().containsKey(new SecondSinceEpocVirtualKey(curSecond)))) {
```

Comment:
Law of Demeter would be nice to apply again here, so that we can e.g. do `schedules.get().hasSchedulesExpiringIn(curSecond)`.